### PR TITLE
ICU-22876 C++ UnicodeSet/USet easy item iteration

### PR DESCRIPTION
**Pull request by markusicu**
Modern C++ iteration over the code points, ranges, or strings, or all of the elements, in a C++ UnicodeSet or C USet.

API approved by ICU-TC on 2024sep05.
Minor change sep09 during implementation testing:
Changed CodePointRange::iterator::operator++() from void to returning iterator & (usual signature)

- UnicodeSet.ranges() returns a USetRanges which has begin() & end() which return USetRangeIterator which works like an idiomatic C++ iterator.
- Parallel for UnicodeSet.codePoints() & UnicodeSet.strings().
- For iterating over all elements (code points & strings) in C++, by popular demand, I added begin() & end() directly to class UnicodeSet, similar to Java UnicodeSet which *is an* Iterable. The header-only wrapper still has separate range and iterator classes.
- For strings, adding new C API to get strings without copying their contents.
- Also, UnicodeSet.strings() wants to be called that in C++ just like in Java, and just like codePoints() & ranges(). That then required me to rename the *field* strings to strings_.
- Operates on C API USet, so that C API users should be able to use it. C++ UnicodeSet.ranges() and it siblings are trivial convenience functions.
- As discussed some time ago, this is written as a header-only API. It is the first of its kind.
- As such, I am defining a U_HEADER_NESTED_NAMESPACE and a U_HEADER_ONLY_NAMESPACE which put the new API into either icu::header when compiling user code or icu::internal when compiling ICU library code, so that even if we instantiate the API inside ICU it won't be the same as what a caller sees.
  - Exception: This does not work on Windows where we explicitly export APIs, because the library does not export the member functions that return the icu::header types.
  - We also don't need this on Windows because we explicitly export APIs.
  - So on Windows the namespace is always icu::header, and the UnicodeSet member functions are exported, but the USet wrapper types are not.
- Naming: I am using USetXyz class names because these classes actually operate on a USet, the C wrapper around a C++ UnicodeSet, and they are defined in unicode/uset.h. This would also allow future additions of UnicodeSetXyz that don't go through the C API, if we ever wanted that.

Sample real, working call sites:
```
    UnicodeSet set(u"[abcçカ🚴]", errorCode);
    for (auto [start, end] : set.ranges()) {
        printf("set.range U+%04lx..U+%04lx\n", (long)start, (long)end);
    }

    UnicodeSet set(u"[abcçカ🚴{}{abc}{de}]", errorCode);
    for (auto el : set) {
        std::string u8;
        printf("set.string length %ld \"%s\"\n", (long)el.length(), el.toUTF8String(u8).c_str());
    }
```

Output:
```
   utility {
      UnicodeSetTest {
         TestUnicodeSetRanges {
set.range U+0061..U+0063
set.range U+00e7..U+00e7
set.range U+30ab..U+30ab
set.range U+1f6b4..U+1f6b4
         } OK:   TestUnicodeSetRanges 
      } OK:   UnicodeSetTest 
   } OK:   utility 

   utility {
      UnicodeSetTest {
         TestUnicodeSetElements {
set.string length 1 "a"
set.string length 1 "b"
set.string length 1 "c"
set.string length 1 "ç"
set.string length 1 "カ"
set.string length 2 "🚴"
set.string length 0 ""
set.string length 3 "abc"
set.string length 2 "de"
         } OK:   TestUnicodeSetElements 
      } OK:   UnicodeSetTest 
   } OK:   utility 
```

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22876
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/unicode/uset.h
Comment (eggrobin) [inline icu4c/source/common/unicode/uset.h:1317]: My immediate thought is that `std::pair<UChar32, UChar32>` is too nondescript a type for a range of code points.
I would expect it to be, well, a [range](https://en.cppreference.com/w/cpp/ranges/range).
In particular, I would expect to be able to iterate over it:
```C++
for (auto const range: set.ranges()) {
  if (MayContainCats(range)) {
    for (UChar32 const cp : range) {
      if (u_iscat(cp)) {
        u_pet(cp);
      }
    }
  }
}
```

There is a bit of boilerplate needed to get that, but I just remembered that conveniently some egg wrote something like it back in October: https://github.com/unicode-org/unicodetools/blob/ce2ff440fca747e32e028594d8484db6e1d33afb/c/uca/sifter/ucd.hpp#L29-L91.

Comment (markusicu) [reply icu4c/source/common/unicode/uset.h:1317]: > My immediate thought is that `std::pair<UChar32, UChar32>` is too nondescript a type for a range of code points. I would expect it to be, well, a [range](https://en.cppreference.com/w/cpp/ranges/range).

Makes sense, and it's tempting, but then either I can't use the nifty `for (auto [start, end] : set.ranges())` loop, or else I have to make funny range field names public. (See below.)

I also don't see why someone would want to do the double loop -- the next thing I am going to add is a code point iterator. Just like in Java, we give easy iteration over either code points or ranges, and callers should choose what's most useful to them. They should not need a double loop.

```
class CodePointRange {
public:
    CodePointRange(UChar32 start, UChar32 end) : start_(start), end_(end) {}
    CodePointRange(const CodePointRange &other) = default;
    inline operator std::pair<UChar32, UChar32>() const { return std::make_pair(start_, end_); }
    UChar32 begin() const { return start_; }
    UChar32 end() const { return end_ + 1; }

private:
    UChar32 start_, end_;
};

usettest.cpp:4274:16: error: cannot decompose private member 'start_' of 'icu_76::header::CodePointRange'
    for (auto [start, end] : set.ranges()) {
```

Comment (eggrobin) [reply icu4c/source/common/unicode/uset.h:1317]: > but then either I can't use the nifty for (auto [start, end] : set.ranges()) loop, or else I have to make funny range field names public. (See below.)

I think _especially_ If you want to allow this nifty syntax, you should use a proper type, and then those member variables can have names and comments to clarify their semantics (iirc ICU and standard C++ have different opinions on what end means?). I see no problem with making it a struct with two public member variables while also giving it the structure of a range.

As for the use case, that double loop was a toy example, but being able to treat it as a range, and to feed it to various things that operate on ranges, seems useful. (certainly having a size() seems convenient).

Comment (eggrobin) [reply icu4c/source/common/unicode/uset.h:1324]: > iirc ICU and standard C++ have different opinions on what end means

Indeed looking at your example above
```C++
    for (auto [start, end] : set.ranges()) {
        printf("set.range U+%04lx..U+%04lx\n", (long)start, (long)end);
    }
```
it looks like those are inclusive ranges, which are not the norm in C++. This is fine, since those _are_ the norm in Unicode, but then they really want to be their own type, not a nondescript pair.
The member variables can probably be called first and last or somesuch to avoid the word `end` which would be interpreted as something else here.

Comment (markusicu) [reply icu4c/source/common/unicode/uset.h]: Ok, done. Added CodePointRange and got it to work.
Note that in ICU, especially in UnicodeSet/USet, ranges have _always_ been inclusive start..end.

**Comment by markusicu** [comment]
Hi everyone:
- I think I have a complete set of new APIs now for this purpose, including the popular "iterate over all elements" thing.
- I have updated the PR description, please read.
- As noted there, I am considering changing the prefixes of the new classes from UnicodeSetXyz to USetXyz, since they are really operating on a C USet. Ok? (One could counter-argue that in C++ we use UnicodeSet. Or that few people will actually type these class names.)
- Please take a look at the sample code in usettest.cpp. (No real unit test code yet.) Looks nice and colloquial?
- Please take a look at the API *signatures*.
- Please take a look at the NOTEs and TODOs.
- @eggrobin I would like to ask you in particular to look at the implementation code as well.

**Comment by markusicu** [comment]
@richgillam do you care whether the prefix of the new classes, which operate on USet but are in C++ and also created by UnicodeSet, is USetXyz or UnicodeSetXyz? For example,
- USetCodePointIterator
- UnicodeSetCodePointIterator

Right now, I am leaning towards the shorter version which seems slightly more accurate.
I should really send the approval email today...

@aheninger opinion?

**Comment by markusicu** [comment]
I used Craig as a sounding board for the class prefix naming question. What tipped him over the top is that the new types are in uset.h, so he would expect them to start with "USet". I like that. Will rename.

**Comment by markusicu** [comment]
Renamed. This would also allow future additions of UnicodeSetXyz that don't go through the C API, if we ever wanted that.

**Comment by richgillam** [comment]
> @richgillam do you care whether the prefix of the new classes, which operate on USet but are in C++ and also created by UnicodeSet, is USetXyz or UnicodeSetXyz? For example,
> 
> * USetCodePointIterator
> * UnicodeSetCodePointIterator

Yes, I like USet here, for the reasons you outline.

File: icu4c/source/common/unicode/uset.h
Comment (richgillam) [inline icu4c/source/common/unicode/uset.h:1337]: I think you can take this comment out now.

Comment (markusicu) [reply icu4c/source/common/unicode/uset.h]: whoops! thanks -- done

**Review by richgillam — APPROVED** [review]
Wow, that's a lot of code to do something that seems like it should be fairly simple!

I'm not sure I completely followed everything, but it all seems to make sense.  But I'll defer to the other reviewers, all of whom understand what's going on here better than I do.

**Comment by markusicu** [comment]
> Wow, that's a lot of code to do something that seems like it should be fairly simple!

I agree!
It's all because C++ didn't create a real iterator pattern and instead stuck with implementing iterators that look and quack like pointers.

File: icu4c/source/common/unicode/uset.h
Comment (eggrobin) [inline icu4c/source/common/unicode/uset.h:1636]: Hm, when we discussed this one in ICU-TC, I was thinking of std::whateverstring, but of course we probably want UnicodeString here.
Does UnicodeString have the small-string optimization though?

File: icu4c/source/common/unicode/uset.h
Comment (eggrobin) [inline icu4c/source/common/unicode/uset.h:1630]: It is common for container modification to invalidate iterators, so I would not worry about this.

File: icu4c/source/common/unicode/uset.h
Comment (markusicu) [reply icu4c/source/common/unicode/uset.h:1636]: > Hm, when we discussed this one in ICU-TC, I was thinking of std::whateverstring, but of course we probably want UnicodeString here.

I chose to return a UnicodeString because it's easy to make one from a code point, and fits well with other ICU APIs.

> Does UnicodeString have the small-string optimization though?

yes

**Comment by markusicu** [comment]
@eggrobin FYI commit ed5e8e30513b8d7d9071047e28681af8abb0ba57 addresses the ICU-TC API proposal review

**Comment by markusicu** [comment]
I think I am all done.
It seems to all work, the API reflects what has been approved, with one minor exception noted in the PR description.
API docs all there, although minimal for boilerplate C++ idioms.
Tested via range-based for loops as well as explicit calls of each new function.

It might still make sense to review this one commit at a time.
Once approved I will squash, then merge.

**Comment by markusicu** [comment]
Dear reviewers, it would be lovely to get this reviewed, feedback addressed, approved, and merged... fairly quickly, to unblock @roubert's follow-up work with the header-only collator predicates. I would be happy to jump into a video meeting to talk things through.

**Review by richgillam — APPROVED** [review]
Looks like all that's changed since the last time I reviewed this is documentation and tests.  The new documentation and tests look good.

File: icu4c/source/common/unicode/uset.h
Comment (markusicu) [reply icu4c/source/common/unicode/uset.h]: done, PTAL

**Review by echeran — APPROVED** [review]
LGTM!

**Review by richgillam — APPROVED** [review]
New documentation also looks good.

**Pull request merged**
